src/Compilers/CSharp/Portable/Parser/LanguageParser_Patterns.cs

```diff
@@ -543,8 +543,14 @@ SyntaxKind.GreaterThanToken or
             // However, such behavior isn't much desirable when parsing pattern of a case label in a switch statement. For instance, consider the following example: `case { Prop: { }: case ...`.
             // Normally we would skip second `:` and `case` keyword after it as bad tokens and continue parsing pattern, which produces a lot of noise errors.
             // In order to avoid that and produce single error of missing `}` we exit on unexpected `:` in such cases.
-            // The same thing can be said about switch expression and `=>` token.
-            if (@this._termState.HasFlag(TerminatorState.IsExpressionOrPatternInLabelOfSwitchStatementOrExpression) && @this.CurrentToken.Kind is SyntaxKind.ColonToken or SyntaxKind.EqualsGreaterThanToken)
+            if (@this._termState.HasFlag(TerminatorState.IsExpressionOrPatternInCaseLabelOfSwitchStatement) && @this.CurrentToken.Kind is SyntaxKind.ColonToken)
+                return PostSkipAction.Abort;
+
+            // This is pretty much the same as above, but for switch expressions and `=>` token.
+            // The reason why we cannot use single flag for both cases is because we only want `:` to be "exit" token for switch statements and `=>` for switch expressions.
+            // Consider the following example: `case (() => 0):`. Normally `=>` is treated as bad separator, so we parse this basically the same as `case ((), 1):`, which is syntactically valid.
```

Comment:
Bleh. I see. I take it that we don't have a fallback path where we simply parse arbitrary expressions here for recovery, and it would maybe be more disruptive to start doing that here.